### ISSUE TITLE
Very slow training when implementing low-data siamese models

### ISSUE THREAD

**Issue by chf42** [deepchem/deepchem#770]
Hi all,

I am trying to implement the siamese model in low-data folder, i.e. "tox_siamese_one_fold.py". The example code works fine, but when I switched to my own single-task data (2000+ samples), it went very slow in the process of "sampling support". It took long time (perhaps one-hour) to get another iteration. I am thinking there must be something wrong, but I can't figure it out.  **Would you mind help me check what might be the reason that caused slow training?**

Thanks,
Cheng

Here is the code: 

n_pos = 1
n_neg = 1
test_batch_size = 39
support_batch_size = n_pos + n_neg
nb_epochs = 1
n_train_trials = 2000
n_eval_trials = 20
n_steps_per_trial = 1
learning_rate = 1e-4
log_every_n_samples = 50

tasks, train_set,test_set,transformers = load_train(featurizer='GraphConv')

n_feat = 75

def model_evaluation (train,test):
	metric = dc.metrics.Metric(dc.metrics.kappa_score, np.mean, mode="classification")
	**support_model = dc.nn.SequentialSupportGraph(n_feat)** (I know it has been deprecated, but it worked well for example code.  So I don't think this is the reason that causes slow training problems )
	support_model.add(dc.nn.GraphConv(64, n_feat, activation='relu'))
	support_model.add(dc.nn.GraphPool())
	support_model.add(dc.nn.GraphConv(128, 64, activation='relu'))
	support_model.add(dc.nn.GraphPool())
	support_model.add(dc.nn.GraphConv(64, 128, activation='relu'))
	support_model.add(dc.nn.GraphPool())
	support_model.add(dc.nn.Dense(128, 64, activation='tanh'))

	support_model.add_test(dc.nn.GraphGather(test_batch_size, activation='tanh'))
	support_model.add_support(
	dc.nn.GraphGather(support_batch_size, activation='tanh'))

	model = dc.models.SupportGraphClassifier(
		support_model,
		test_batch_size=test_batch_size,
		support_batch_size=support_batch_size,
		learning_rate=learning_rate)
	model.fit(
		train,
		nb_epochs=nb_epochs,
		n_episodes_per_epoch=n_train_trials,
		n_pos=n_pos,
		n_neg=n_neg,
		log_every_n_samples=log_every_n_samples)
	mean_scores, std_scores = model.evaluate(
	test, metric, n_pos, n_neg, n_trials=n_eval_trials)
	return mean_scores,std_scores

value_test,sed_scores = model_evaluation (train_set,test_set)
print("Test Validation")
print value_test

**Comment by rbharath** [comment]
Which part is slow? (which line)

**Comment by chf42** [comment]
@rbharath 
I think it corresponds to the support layer: 
**support_model.add_support(
dc.nn.GraphGather(support_batch_size, activation='tanh'))

model = dc.models.SupportGraphClassifier(
	support_model,
	test_batch_size=test_batch_size,
	support_batch_size=support_batch_size,
	learning_rate=learning_rate)
model.fit(
	train,
	nb_epochs=nb_epochs,
	n_episodes_per_epoch=n_train_trials,
	n_pos=n_pos,
	n_neg=n_neg,
	log_every_n_samples=log_every_n_samples)**

I put output as follows: 

Loading shard 1 of size 8192.
Featurizing sample 0
TIMING: featurizing shard 0 took 0.767 s
TIMING: dataset construction took 1.547 s
Loading dataset from disk.
TIMING: dataset construction took 9.630 s
Loading dataset from disk.

**Sampling support 0
Sampling support 50
Sampling support 100
Sampling support 150**

This part is quite slow, which I think corresponds to the support layer.

Thanks,
Cheng

**Comment by rbharath** [comment]
I'm not sure why it's slow. I'm also hesitant to put too much time into debugging it, since the code is deprecated.

I'll work on getting a TensorGraph version of the one shot models functional so we have a clean slate

**Comment by rbharath** [comment]
Closing this issue since it refers to the old implementation (non TensorGraph). The TensorGraph implementation is being worked on https://github.com/deepchem/deepchem/pull/1084. Let's shift discussion of implementation issues there.
